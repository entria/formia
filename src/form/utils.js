import get from 'lodash.get';
import set from 'lodash.set';
import isEqual from 'lodash.isequal';

export const handleInitialValuesUpdate = (
  currentValues = {},
  currentInitialValues = {},
  newInitialValues = {},
) => {
  const updatedValues = { ...currentValues };

  Object.keys(newInitialValues).map(fieldName => {
    const fieldCurrentValue = get(currentValues, fieldName, null);
    const fieldCurrentInitialValue = get(currentInitialValues, fieldName, null);
    const isFieldPristine = isEqual(fieldCurrentValue, fieldCurrentInitialValue);

    if (!isFieldPristine) {
      return;
    }

    const fieldNewInitialValue = get(newInitialValues, fieldName, null);
    set(updatedValues, fieldName, fieldNewInitialValue);
  });

  return updatedValues;
};
