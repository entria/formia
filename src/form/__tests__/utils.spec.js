import { handleInitialValuesUpdate } from '../utils';

it('handleInitialValuesUpdate - default props', () => {
  expect(handleInitialValuesUpdate()).toMatchSnapshot();
});

it('handleInitialValuesUpdate - async loading', () => {
  const currentValues = {};
  const currentInitialValues = {};
  const newInitialValues = { name: 'Barney Stinson' };

  expect(
    handleInitialValuesUpdate(currentValues, currentInitialValues, newInitialValues),
  ).toMatchSnapshot();
});

it('handleInitialValuesUpdate - async update if field is pristine', () => {
  const currentValues = { name: 'Barney Stinson' };
  const currentInitialValues = { name: 'Barney Stinson' };
  const newInitialValues = { name: 'Barney Stinson Awesome' };

  expect(
    handleInitialValuesUpdate(currentValues, currentInitialValues, newInitialValues),
  ).toMatchSnapshot();
});

it('handleInitialValuesUpdate - async update if field is dirty', () => {
  const currentValues = { name: 'Legendary Barney Stinson' };
  const currentInitialValues = { name: 'Barney Stinson' };
  const newInitialValues = { name: 'Barney Stinson Awesome' };

  expect(
    handleInitialValuesUpdate(currentValues, currentInitialValues, newInitialValues),
  ).toMatchSnapshot();
});
