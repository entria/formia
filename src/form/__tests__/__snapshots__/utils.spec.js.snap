// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleInitialValuesUpdate - async loading 1`] = `
Object {
  "name": "Barney Stinson",
}
`;

exports[`handleInitialValuesUpdate - async update if field is dirty 1`] = `
Object {
  "name": "Legendary Barney Stinson",
}
`;

exports[`handleInitialValuesUpdate - async update if field is pristine 1`] = `
Object {
  "name": "Barney Stinson Awesome",
}
`;

exports[`handleInitialValuesUpdate - default props 1`] = `Object {}`;
